app/scripts/controllers/authentication/services.ts

```diff
@@ -0,0 +1,96 @@
+const AUTH_ENDPOINT = process.env.AUTH_API || '';
+export const AUTH_NONCE_ENDPOINT = `${AUTH_ENDPOINT}/api/v2/nonce`;
+export const AUTH_LOGIN_ENDPOINT = `${AUTH_ENDPOINT}/api/v2/snaps/login`;
+
+const OIDC_ENDPOINT = process.env.OIDC_API || '';
+export const OIDC_TOKENS_ENDPOINT = `${OIDC_ENDPOINT}/oauth2/token`;
+const OIDC_CLIENT_ID = process.env.OIDC_CLIENT_ID || '';
+const OIDC_GRANT_TYPE = process.env.OIDC_GRANT_TYPE || '';
+
+export type NonceResponse = {
+  nonce: string;
+};
+export async function getNonce(publicKey: string): Promise<string | null> {
+  const nonceUrl = new URL(AUTH_NONCE_ENDPOINT);
+  nonceUrl.searchParams.set('identifier', publicKey);
+
+  const nonce = await fetch(nonceUrl.toString())
+    .then((r) => {
+      return r.ok ? r.json() : null;
+    })
+    .then((r: NonceResponse | null) => r?.nonce)
+    .catch(() => null);
```

Comment:
should it be error catching here? if we don't catch, IMO there's no need for return null

Reply:
I'm not a fan of error propagation in JS.
The consumer of this function will then need to handle: string return; a null return; and a hidden potentially thrown error.

I'd rather catch as early as possible and either treat the result as a value (in this case `null`) or make it into an "option" type (e.g. `{ isValid: false, error: e }`).

In this case I'm happy to opt for resolving catches to `null` as there is not information I care about propagating up to the consumer (but might change in future).

Reply:
More to the point, if we are using the 'await' terminology here why are we using .then and .catch chaining? 

Is it to avoid intermediary naming? 

Aka
```
let nonce = null;
try {
  const nonceResponse = await fetch(nonceUrl.toString());
  nonce = nonceResponse.ok ? r.json()?.nonce : null;
} catch () {
  // possible logLevel or some sort of sentryMessage so that we're aware this is happening?
}
```

OR drop the await.

Reply:
> Do we not allow mixing of promises and await syntaxes?
(personal) I prefer .then chaining to keep sections (e.g. transformations) scoped. Likewise it allows error handling immediately than a global try-catch where you need to differentiate errors.

Mixing these styles is confusing, personally, to me, and we should be clear here that the goal of pull request reviews is to get code into a state where someone who isn't you can maintain that code afterward. I'd love to hear other opinions here and it isn't just my view versus yours its a collective. 

> (personal) I prefer .then chaining to keep sections (e.g. transformations) scoped.
What do you mean here by "scoped transformations"? 

> Likewise it allows error handling immediately than a global try-catch where you need to differentiate errors.

I'm not sure what you mean by a global try catch. The try catch in my example is scoped around the same code that the chain method covers.

Reply:
Updated comment see snippet below.

These are all personal coding decisions, and yep we can leave it to a collective. I however don't think we should dissuade patterns that may be handy (such as mixing async/await with promises) otherwise we might as well not adopt any modern syntax (any TS syntax or shorthands, what about future JS syntax like the pipe operator?).

If there are patterns we as a collective do agree not to use (e.g. Typescript Constructor Shorthand), then lets use concrete this through ESLint rules, otherwise keep it flexible if there are not logical issues.

I do agree that the above code was pretty terse (especially the `.then(r => r.ok ? null : r.json())`), and have opted for an async/await.

```ts
// Treat errors as values.
// Think similar to an Option type, Result type, or even a null error result.
const res1: Foo | null = await foo().then(/* do stuff*/).catch(() => null)
if (!res1) {
  // handle case here
}

const res2: Bar | null = await bar().then(/* do stuff*/).catch(() => null)
if (!res2) {
  // handle case here
}

// VS try-catch clauses
try {
  const res1: Foo = await foo()
  const res2: Bar = await bar()
} catch (e: unknown) {
  // error here is unknown, we need to somehow instanceof or "figure out" which error is which
}

// OR let-try-catch
// this is okay, but why have a mutable variable that is only set once?
let res1: Foo | undefined;
try {
  res1 = await foo()
} catch (e: unknown) {
  // handle error
}

let res2: Bar | undefined;
try {
  res2 = await bar()
} catch (e: unknown) {
  // handle error
}
```

Reply:
> I'm not a fan of hidden error propagation in JS/TS.

I believe there's a misunderstanding there. IMO return `null` does not help with the error catching, we should throw error out when the network request is failed. 

```
try {
  const res1: Foo = await foo()
  const res2: Bar = await bar()
} catch (e: unknown) {
  // error here is unknown, we need to somehow instanceof or "figure out" which error is which
}
```
I personally found this solution acceptable, as we actually don't want to fire 2nd request when 1st one is failed, and the error will be caught clearly. However, the `e: unknown` is a question, i believe the `foo()` will return the `error` object as we expect.

Reply:
> IMO return null does not help with the error catching
This is valid, but again personal. IMO if I wanted to handle the error I would usually "return" some error instead of throwing. See how these [services are used](https://github.com/MetaMask/metamask-extension/pull/23286/files#diff-2061685b6ae9e88df78ce92c7c0a5fcdaadf5f4b7fd09f8ce8e66aab6eb62c38R172) , we handle null's here and then can throw.

> However, the e: unknown is a question, i believe the foo() will return the error object as we expect
Close, TS actually infers the catch as an `unknown` - [see playground](https://www.typescriptlang.org/play?#code/C4TwDgpgBAYg9nKBeKBvKBDAXFAzsAJwEsA7AcygF8AoDXEEgYygDMBXJ4IuE1hACgCUOAAoE4AWyK4IAHnhwAfGmpQowABbiA7lBIRdAUQLiC-AOQsBgqAFooEuIwDWUCCbgFzg6jVr0mVg5GLh51CHwhFTVCEGi1KEYefCgCCJwFZExtDCJgPjghVSpEjGBGDSh+CBtUYrUAegaEtQA9AH5qeqgmqAAjCEYMNhk3DwIoDXdoaSgOZxI4bRIAGihtaAg8qYn9CAATdURQSExcGQJ8vKhPOdHSfAwmCDgWRKmXXAA6bqI36qgD2AT0YLzexlMtW6aiSJBS7lMADkMB4NocUBAANzQnrNBIdbp+Fq9ACSEgwZFI0BAcDYE1i-QANk5XBoMIdySQ4vs-ixpiR8nQGMwhozGd9uqS3jS2OsnsADkcHE8KdBNNAeSw+WkBWNTKUZLg1jL1rTGYc9odgIg6Bd8rcKoNXBAMBUDRAfsTmgBBMi5XiaWazFhUgCEUBE7lwPAwjJuYFI3F4swAViN8mySPtGWqpnrbnRAfkkhIIqkyjt1Jn1AQQA0huUNJ6qL4gA).
You will either need to type assert it to the error or use an `instanceof` checks.

Again this is fine, and we are bike shedding since any solution works. Everyone has their own preference:
- Throw errors vs return errors
- Handle errors inline vs handle all errors on catch statement.

I've updated the code to use async await, but am still not throwing as there is no point.

Reply:
@Prithpal-Sooriya everyone making their own personal coding decisions in terms of patterns will lead us down a path of chaos, which is why we are having these conversations -- anything that is not understood by the code reviewers is a clear indicator of an issue with maintainability going forward. We are pushing towards standards and guidelines so that everyone's code can mesh effectively and so that the next person after you can effectively maintain it. That isn't bike shedding. These are important questions for the health of the code base.

Reply:
I like playing devils advocate here 😄 

Yep I **_mostly_** agree, but until these rules are concrete & we add linting rules, code should be flexible.
If code is documented, tested, and functions are short & clear I see no reason for these to be blockers.

The original/main comment:
> should it be error catching here? if we don't catch, **IMO** there's no need for return null

Is a discussion about error catching and if we should return null. We are all writing JavaScript & Typescript, there are probably many different error handling patterns you can create (now and in the future). Key thing here is that this is all opinions, we can even raise this see what others consider.

See the function definition, the return type, the tests and where/how it is used - I think it is completely fine to:
1. Catch any errors to prevent any unhandled errors propagating up.
2. Return null if the endpoint fails.

Is it fine to return null? I think so! 
Lets look at this function definition `getNonce(): Promise<string | null>`. As a consumer of this, I can see that this will may give me a nonce, or null. If it is a null, I can choose what I want to do (now throw an error, or even go through a different flow).